### ISSUE TITLE
Create User - Row was updated or deleted by another transaction

### ISSUE THREAD

**Issue by fairking** [nhibernate/NHibernate.AspNetCore.Identity#17]
I got the following error when a new user is created the second time:
```
System.AggregateException: One or more errors occurred. (Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect)
 at NHibernate.Impl.SessionImpl.FlushAsync(CancellationToken cancellationToken)
   at NHibernate.AspNetCore.Identity.UserStore`2.FlushChangesAsync(CancellationToken cancellationToken) in C:\Code\NHibernate.AspNetCore.Identity\src\NHibernate.AspNetCore.Identity\UserStore.cs:line 587
   at NHibernate.AspNetCore.Identity.UserStore`2.CreateAsync(TUser user, CancellationToken cancellationToken) in C:\Code\NHibernate.AspNetCore.Identity\src\NHibernate.AspNetCore.Identity\UserStore.cs:line 47
   at Microsoft.AspNetCore.Identity.UserManager`1.CreateAsync(TUser user)
   at Microsoft.AspNetCore.Identity.UserManager`1.CreateAsync(TUser user, String password)
```

My implementation as follows:
```
...
cfg.SetProperty("default_flush_mode", "Commit");
cfg.SetProperty("connection.isolation", "ReadUncommitted");
cfg.SetProperty("adonet.batch_size", "10");
cfg.SetProperty("default_batch_fetch_size", "1");
cfg.SetProperty("generate_statistics", "false");
cfg.SetProperty("show_sql", "false");
cfg.SetProperty("format_sql", "false");
cfg.SetProperty("connection.driver_class", "NHibernate.Driver.SqlClientDriver,NHibernate");
cfg.SetProperty("dialect", "NHibernate.Dialect.MsSql2008Dialect");
...
services.AddSingleton(s => s.GetRequiredService<Configuration>().BuildSessionFactory());
services.AddScoped(s => s.GetRequiredService<ISessionFactory>().WithOptions().Interceptor(new AppInterceptor(s)).OpenSession());
...
services.AddIdentityCore<AppUser>(options => { ... })
    .AddRoles<AppRole>()
    .AddHibernateStores()
    .AddDefaultTokenProviders();
services.AddScoped<IRoleStore<AppRole>>(s => new RoleStore<AppRole>(s.GetRequiredService<ISession>(), new IdentityErrorDescriber()));
...
using (var scope = host.Services.CreateScope())
{
    using (var tran = session.BeginTransaction())
    {
        var appUser1 = new AppUser(name1, email1);
        await _userManager.CreateUserAsync(appUser1, roles, password);
        var appUser2 = new AppUser(name2, email2);
        await _userManager.CreateUserAsync(appUser2, roles, password); // Error is thrown here
        tran.Commit();
    }
}
```
As u can see the error is thrown on the second user creation `_userManager.CreateUserAsync(appUser2, roles, password)`. 
So I assume it is because of the `session.Clear()` in `UserStore.FlushChangesAsync()`.

**Comment by fairking** [comment]
As a solution there could be the following change I guess:
![image](https://user-images.githubusercontent.com/13495631/72756483-432f7080-3bc5-11ea-949e-67302b75387b.png)

What do you think guys?

**Comment by beginor** [comment]
The `CreateUserAsync` method returns a `Task` instance, you should wait for it's completion, the code should like this:

```c#
var task = _userManager.CreateUserAsync(appUser1, roles, password);
task.Wait();
task = _userManager.CreateUserAsync(appUser2, roles, password);
task.Wait();
```

Or using the `async/await` pattern:

```c#
await _userManager.CreateUserAsync(appUser1, roles, password);
await _userManager.CreateUserAsync(appUser2, roles, password);
```

**Comment by fairking** [comment]
Sorry @beginor . It is my mistake in this example. I fixed it. There is `await` in my actual application. Some of the lines I manually typed to simplify the example.

**Comment by beginor** [comment]
@fairking I think this is OK by design.

The `UserManager<AppUser>` uses a `UserStore<AppUser>` to save users to database, and the `UserStore<AppUser>` uses an `ISession` from `IServiceProvider` , so there are two session instances in your code above.

Usually we do not need to create more then one users in a transaction, but if you want to create more then one users in a custom transaction, I think you can create a `UserStore<AppUser>` by hand like this:

```c#
var sessionFactory = cfg.BuildSessionFactory();
var session = sessionFactory.OpenSession();
var transaction = session.BeginTransaction();
store = new RoleStore<NHIdentityRole>(session, new IdentityErrorDescriber());
```

**Comment by fairking** [comment]
Hi @beginor I could not get it. You said: "UserStore<AppUser> uses an ISession from IServiceProvider , so there are two session instances in your code above";
But I can see only one scope, so it must be the same session. There are no two sessions. Two users created with the same single session.
You might be thinking there is another `asp.net` session (scope) along with the scope I create in my example, but it is not. My example seeds the database with data before the app starts so there is not running any other session.
Anyway, I will look for a different solution then. Thanks for your help.

**Comment by fairking** [comment]
Ok. I attached the `NHibernate.AspNetCore.Identity` locally to my solution and removed the `session.Clear` completely.
I cannot split two user creation to two different transactions. It is part of one transaction in my business logic. I got transactions and cannot allow for session clear in the middle of the transaction. I clear sessions at the end of the transaction by myself. By the way thanks for your help.

**Comment by beginor** [comment]
@fairking  I am sure that there are more then one nhibernate session in your code because asp.net core will create a separate scope for each request. 

And if you want to create more then one users with a single nhibernate session with in transaction please refer to the following code:

```c#
[HttpGet("batch-create")]
public async Task<ActionResult> BatchCreate() {
    var provider = HttpContext.RequestServices;
    using var scope = provider.CreateScope();
    var session = provider.GetService<ISession>();
    var manager = provider.GetService<UserManager<AppUser>>();
    using var tx = session.BeginTransaction();
    try {
        var user1 = new AppUser {
            UserName = "user1",
            Email = "user1@nonexit.com"
        };
        await manager.CreateAsync(user1);
        var user2 = new AppUser {
            UserName = "user2",
            Email = "user2@nonexit.com"
        };
        await manager.CreateAsync(user2);
        await tx.CommitAsync();
        return Ok();
    }
    catch (Exception ex) {
        await tx.RollbackAsync();
        return StatusCode(500);
    }
}
```

**Comment by fairking** [comment]
Thanks for your help @beginor ,
I create those users before asp.net core app starts. Bellow is my code:
![image](https://user-images.githubusercontent.com/13495631/73124308-eb608300-3f91-11ea-9fe2-0c07c68da779.png)

![image](https://user-images.githubusercontent.com/13495631/73124390-c3bdea80-3f92-11ea-9c62-db45f592b27f.png)

At the momet I got removed `session.Clear()` in `NHibernate.AspNetCore.Identity` and everything works fine. I could not understand why do we need to clear the session when the user creation could be a part of a bigger business logic. Eg. Create user, create default user settings, apply department.

Is it related to some security matters?
